js/modules/loki_snode_api.js

```diff
@@ -18,6 +18,226 @@ class LokiSnodeAPI {
+      params,
+    };
+
+    const fetchOptions = {
+      method: 'POST',
+      body: JSON.stringify(body),
+      headers: { 'Content-Type': 'application/json' },
+      timeout: 10000, // 10s, we want a smaller timeout for testing
+    };
+
+    process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
+
+    let response;
+
+    try {
+      response = await nodeFetch(url, fetchOptions);
+    } catch (e) {
+      if (e.type === 'request-timeout') {
+        log.warn(`test timeout for node,`, snode);
+      }
+      return false;
+    } finally {
+      process.env.NODE_TLS_REJECT_UNAUTHORIZED = '1';
+    }
+
+    if (!response.ok) {
+      log.info(`Node failed the guard test:`, snode);
+    }
+
+    return response.ok;
+  }
+
+  async selectGuardNodes() {
+    const _ = window.Lodash;
+
+    const nodePool = await this.getRandomSnodePool();
+
+    if (nodePool.length === 0) {
+      log.error(`Could not select guarn nodes: node pool is empty`);
+      return [];
+    }
+
+    const shuffled = _.shuffle(nodePool);
+
+    let guardNodes = [];
+
+    const DESIRED_GUARD_COUNT = 3;
+
+    // The use of await inside while is intentional:
+    // we only want to repeat if the await fails
+    // eslint-disable-next-line-no-await-in-loop
+    while (guardNodes.length < 3) {
+      if (shuffled.length < DESIRED_GUARD_COUNT) {
+        log.error(`Not enought nodes in the pool`);
+        break;
+      }
+
+      const candidateNodes = shuffled.splice(0, DESIRED_GUARD_COUNT);
+
+      // Test all three nodes at once
+      // eslint-disable-next-line no-await-in-loop
+      const idxOk = await Promise.all(
+        candidateNodes.map(n => this.testGuardNode(n))
+      );
+
+      const goodNodes = _.zip(idxOk, candidateNodes)
+        .filter(x => x[0])
+        .map(x => x[1]);
+
+      guardNodes = _.concat(guardNodes, goodNodes);
+    }
+
+    if (guardNodes.length < DESIRED_GUARD_COUNT) {
+      log.error(
+        `COULD NOT get enough guard nodes, only have: ${guardNodes.length}`
+      );
+    }
+
+    log.info('new guard nodes: ', guardNodes);
+
+    const edKeys = guardNodes.map(n => n.pubkey_ed25519);
+
+    await window.libloki.storage.updateGuardNodes(edKeys);
+
+    return guardNodes;
+  }
+
+  async getOnionPath(toExclude = null) {
+    const _ = window.Lodash;
+
+    const goodPaths = this.onionPaths.filter(x => !x.bad);
+
+    if (goodPaths.length < 2) {
+      log.error(
+        `Must have at least 2 good onion paths, actual: ${goodPaths.length}`
+      );
+      await this.buildNewOnionPaths();
+    }
+
+    const paths = _.shuffle(goodPaths);
+
+    if (!toExclude) {
+      return paths[0];
+    }
+
+    // Select a path that doesn't contain `toExclude`
+    const otherPaths = paths.filter(
+      path =>
+        !_.some(path, node => node.pubkey_ed25519 === toExclude.pubkey_ed25519)
+    );
+
+    if (otherPaths.length === 0) {
+      // This should never happen!
+      log.error('No onion paths available after filtering');
```

Comment:
We should still do a graceful fallback even if this case shouldn't happen.
Otherwise the line below will throw an error, unless that is intended.

Reply:
Yeah, I thought I'd just let whatever uses it throw. I changed it to explicitly throw.